### PR TITLE
Add e2e test for services with no endpoints

### LINKED ISSUE #20767: e2e needed: kube-proxy REJECT rules
#19576 introduced somethng that really should be e2e tested.

### PR DESCRIPTION
**Pull request by vllry**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR adds an e2e test for services that have no endpoints. Expected behaviour is a rejected connection.

master is currently broken, with requests timing out rather than being rejected. There is a fix for the regression in #72534

**Which issue(s) this PR fixes**:

Fixes #20767

**Special notes for your reviewer**:
This test loops until it receives a success (fail, it shouldn't), receives the expected failure, or times out.

* I'm not certain if `framework.KubeProxyEndpointLagTimeout` is the correct timeout to use.
* Is wget string-searching acceptable? Should it be strengthened?
* I opted to print the received error each loop for clarity, but this might be too noisy.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Review by danwinship — CHANGES_REQUESTED** [review]
There's something a bit messed up with your branch, git-wise. (Look in the "Commits" tab here; it shows that your branch includes a merge from PR 1.) You should squash everything together into a single commit, and then force-push that.

> * I'm not certain if `framework.KubeProxyEndpointLagTimeout` is the correct timeout to use.

Yes, that's correct for "retry until we're sure that kube-proxy should have updated things to be the way we expect".

> * Is wget string-searching acceptable? Should it be strengthened?

Ugh. It's not great...

It looks like busybox's wget exits with status 1 on basically any error, so we can't check that either. (And "real" wget is no better; it exits with status 4 for either "connection refused" or "timed out".)

The execpod stuff does at least use a specific version of the busybox image, so it's not like we have to worry about the binary suddenly changing... maybe this is OK?

> * I opted to print the received error each loop for clarity, but this might be too noisy.

That's fine; in the usual case it will only run once. And no one ever looks at the output unless the test fails anyway...

**Comment by vllry** [comment]
Thanks for reviewing @danwinship, I'll get changes up later today.

Regarding detecting wget's error, perhaps I could at least do a better string-match? My thought was that if it fails due to command output changing, that should at least be very visible on retrospect. Technically we could get fancy with writing a specific binary or something to avoid string parsing... but I doubt it's worth it.

**Comment by danwinship** [comment]
/ok-to-test
(though of course we expect the test to fail since #72534 hasn't merged yet)

**Comment by vllry** [comment]
Sorry, I squashed my history but apparently there's an issue stemming from merging k/k into my branch.

**Comment by vllry** [comment]
/remove-sig api-machinery apps cli cluster-lifecycle instrumentation node release scheduling storage
/remove-area kubeadm kubectl kubelet

**Comment by vllry** [comment]
^ Test failures are just this test timing out (as expected before #72534 merges)

**Comment by danwinship** [comment]
> Regarding detecting wget's error, perhaps I could at least do a better string-match?

Better how? I think what you've got is about as good as it's going to get using this approach.

> Technically we could get fancy with writing a specific binary or something to avoid string parsing... but I doubt it's worth it.

And if we did that, we'd have to get that binary added to one of the container images used by the test suite, since we're specifically testing "connecting from a pod" here.

/lgtm
but @thockin might be more dubious about it...
/assign @thockin

File: test/e2e/network/service.go [hunk 1]
```diff
@@ -1748,6 +1749,62 @@ var _ = SIGDescribe("Services", func() {
...
+			TargetPort: intstr.FromInt(80),
+		}}
+
+		By("creating a service with no endpoints")
+		svc, err := jig.CreateServiceWithServicePort(labels, namespace, ports)
+		if err != nil {
+			framework.Failf("Failed to create service: %v", err)
+		}
+
+		port := strconv.Itoa(int(svc.Spec.Ports[0].Port))
```
Comment (thockin) [inline test/e2e/network/service.go]: How could the anything but the "80" you used above?

File: test/e2e/network/service.go [hunk 2]
```diff
@@ -1748,6 +1749,62 @@ var _ = SIGDescribe("Services", func() {
...
+
+		By(fmt.Sprintf("creating %v on node %v", podName, nodeName))
+		execPodName := framework.CreateExecPodOrFail(f.ClientSet, namespace, podName, func(pod *v1.Pod) {
+			pod.Spec.NodeName = nodeName
+		})
+		execPod, err := f.ClientSet.CoreV1().Pods(namespace).Get(execPodName, metav1.GetOptions{})
+		framework.ExpectNoError(err)
+
+		framework.Logf("waiting up to %v wget %v", framework.KubeProxyEndpointLagTimeout, servicePort)
+		cmd := fmt.Sprintf(`wget -T 10 -qO- %v`, servicePort)
```
Comment (thockin) [inline test/e2e/network/service.go]: 10 seems very long if there is a regression.  Why not 2 or 3?

**Review by thockin — COMMENTED** [review]
A couple REALLY minor things, otherwise LGTM

**Comment by vllry** [comment]
/retest

Looks like this was hit with the github API token issue earlier

**Comment by vllry** [comment]
@thockin I've made the suggested cleanup, sorry about the blips 😬

**Review by thockin — COMMENTED** [review]
/lgtm
/approve

THANKS!

**Comment by spiffxp** [comment]
/retest
This appears to be hitting https://github.com/kubernetes/kubernetes/issues/73444

No idea why the other jobs are hitting merge conflicts?

**Pull request merged**